mlir/lib/Driver/CompilerDriver.cpp

```diff
@@ -443,40 +463,54 @@ LogicalResult runLowering(const CompilerOptions &options, MLIRContext *ctx, Modu
                           CompilerOutput &output)
 
 {
+    using std::pair;
+    using Timer = catalyst::utils::Timer<std::ostream>;
+    using LineCounter = catalyst::utils::LinesCount<std::ostream>;
     auto &outputs = output.pipelineOutputs;
     auto pm = PassManager::on<ModuleOp>(ctx, PassManager::Nesting::Implicit);
 
+    if (options.verbosity >= Verbosity::Timing) {
+        auto tm = std::make_unique<DefaultTimingManager>();
+        tm->setOutput(options.diagnosticStream);
+        tm->setEnabled(true);
+        pm.enableTiming(std::move(tm));
+    }
+
     // Maps a pass to zero or one pipelines ended by this pass
     // Maps a pass to its owning pipeline
-    std::unordered_map<const Pass *, Pipeline::Name> pipelineTailMarkers;
-    std::unordered_map<const Pass *, Pipeline::Name> passPipelineNames;
+    std::unordered_map<const Pass *, pair<Pipeline::Name, size_t>> pipelineTailMarkers;
+    std::unordered_map<const Pass *, pair<Pipeline::Name, size_t>> passPipelineNames;
 
     // Fill all the pipe-to-pipeline mappings
-    for (const auto &pipeline : options.pipelinesCfg) {
-        size_t existingPasses = pm.size();
-        if (failed(parsePassPipeline(joinPasses(pipeline.passes), pm, options.diagnosticStream))) {
-            return failure();
-        }
-        if (existingPasses != pm.size()) {
-            const Pass *pass = nullptr;
-            for (size_t pn = existingPasses; pn < pm.size(); pn++) {
-                pass = &(*(pm.begin() + pn));
-                passPipelineNames[pass] = pipeline.name;
+    {
+        size_t pipelineIdx = 0;
+        for (const auto &pipeline : options.pipelinesCfg) {
+            size_t existingPasses = pm.size();
+            if (failed(
+                    parsePassPipeline(joinPasses(pipeline.passes), pm, options.diagnosticStream))) {
+                return failure();
             }
-            assert(pass != nullptr);
-            pipelineTailMarkers[pass] = pipeline.name;
+            if (existingPasses != pm.size()) {
+                const Pass *pass = nullptr;
+                for (size_t pn = existingPasses; pn < pm.size(); pn++) {
+                    pass = &(*(pm.begin() + pn));
+                    passPipelineNames[pass] = pair(pipeline.name, pipelineIdx);
+                }
+                assert(pass != nullptr);
+                pipelineTailMarkers[pass] = pair(pipeline.name, pipelineIdx);
+            }
```

Comment:
Can you explain this bit of code?

Reply:
Oh, this code contained a problem even in the previous version as can be seen from the comment. I have simplified it.

MLIR knows nothing about "pipelines" in our sense. We build a single mlir pipeline with all the passes we provide to it. In order to maintain our own dumping logic (we wanted dumps to have names of our pipelines), we need to do the bookkeeping by ourselves. Here we record pipeline for each pass.

A minor problem here is the following one: one pass might 'end' more than one pipeline if we have empty pipelines. So when we create dumps, we might need to emit some equal dumps with different names to be correct and this was my original proposal. Probably later someone misunderstood and simplified it.

In this PR I would like to avoid working on it.

Reply:
> A minor problem here is the following one: one pass might 'end' more than one pipeline if we have empty pipelines.

I don't think we should worry about empty pipelines. That can be asserted from the frontend as a case that does not happen.

Reply:
I agree it is a minor issue. The reason why I proposed this initially is because I consider a distant-future possibility of automatically-generated pipelines. In this case we will need a design that would accept any input and hold invariant "user can ask a report for _any_ pipeline". FYI I don't plan to insist on this now.